src/transform/classify.rs

```diff
@@ -34,56 +65,194 @@ fn classify_one(mut event: Event, classes: &Vec<(String, Regex)>) -> Event {
+            return _pick_highest_ranking_category(acc, &item);
+        })
+        .clone()
+}
+
+fn _cat_format_to_vec(cat: String) -> Vec<String> {
+    cat.split("->").map(|s| s.trim().into()).collect()
+}
+
+fn _cat_vec_to_format(cat: Vec<String>) -> String {
+    cat.join(" -> ")
+}
+
+#[test]
+fn test_categorize() {
+    let mut e = Event::default();
+    e.data
+        .insert("test".into(), serde_json::json!("just a test"));
+
+    let mut events = vec![e];
+    let cats: Vec<(Vec<String>, Regex)> = vec![
+        (vec!["Test".into()], Regex::new(r"test").unwrap()),
+        (
+            vec!["Test".into(), "Subtest".into()],
+            Regex::new(r"test").unwrap(),
+        ),
+        (vec!["Other".into()], Regex::new(r"nonmatching").unwrap()),
+    ];
+    events = categorize(events, &cats);
+
+    assert_eq!(events.len(), 1);
+    assert_eq!(
+        events.first().unwrap().data.get("$category").unwrap(),
+        &serde_json::json!(vec!["Test", "Subtest"])
+    );
+}
+
+#[test]
+fn test_autotag() {
+    let mut e = Event::default();
+    e.data
+        .insert("test".into(), serde_json::json!("just a test"));
+
+    let mut events = vec![e];
+    let rules: Vec<(String, Regex)> = vec![
+        ("test".into(), Regex::new(r"test").unwrap()),
+        ("test-2".into(), Regex::new(r"test").unwrap()),
+        ("nonmatching".into(), Regex::new(r"nonmatching").unwrap()),
+    ];
+    events = classify(events, &rules);
+
+    assert_eq!(events.len(), 1);
+    assert_eq!(
+        events
+            .first()
+            .unwrap()
+            .data
+            .get("$tags")
+            .unwrap()
+            .as_array()
+            .unwrap()
+            .len(),
+        2
+    );
 }
 
 #[test]
 fn test_classify() {
     let mut e = Event::default();
-    e.data.insert("test".into(), serde_json::json!("just a test"));
+    e.data
+        .insert("test".into(), serde_json::json!("just a test"));
 
-    let events = vec!(e);
-    let classes: Vec<(String, Regex)> = vec!(
+    let mut events = vec![e];
+    let classes: Vec<(String, Regex)> = vec![
         ("#test-tag".into(), Regex::new(r"test").unwrap()),
         ("Test".into(), Regex::new(r"test").unwrap()),
         ("Test -> Subtest".into(), Regex::new(r"test").unwrap()),
         ("Other".into(), Regex::new(r"nonmatching").unwrap()),
-    );
-    let events_classified = classify(events, &classes);
+    ];
+    events = classify(events, &classes);
 
-    assert_eq!(events_classified.len(), 1);
-    assert_eq!(events_classified.first().unwrap().data.get("$tags").unwrap().as_array().unwrap().len(), 3);
-    assert_eq!(events_classified.first().unwrap().data.get("$category").unwrap(), &serde_json::json!("Test -> Subtest"));
+    assert_eq!(events.len(), 1);
+    assert_eq!(
+        events
+            .first()
+            .unwrap()
+            .data
+            .get("$tags")
+            .unwrap()
+            .as_array()
+            .unwrap()
+            .len(),
```

Comment:
Lol, that's one way to assert I guess.
Would be more readable with something like

```
let event = events.first.unwrap();
let tags = event.get("$tags").unwrap();
assert_eq!(tags.as_arrat().unwrap().len(), 3);
```